src/resmom/requests.c

```diff
@@ -1622,12 +1626,166 @@ post_resume(job *pjob, int err)
+	 * completes, or a hard error is returned, or 
+	 * alps_confirm_switch_timeout is reached.
+	 * alps_confirm_switch_timeout is set by default to 
+	 * ALPS_CONF_SWITCH_TIMEOUT 
+	 * NOTE:  The MOM, server and scheduler are blocked while we poll
+	 * ALPS and wait for the SWITCH.
+	 */
+	first_status = 1;
+	first_sleep = 1;
+	begin_time = time(NULL);
+	end_time = begin_time;
+	i = 0;
+	while ((total_time = end_time - begin_time) < timeout_val) {
+		i++;
+		if ((rc = alps_confirm_suspend_resume(pjob, action)) <= 0)
+			break;
+		if (rc == 2) {
+			/* we got a response of "EMPTY" */
+			if (first_status) {
+				/* 
+				 * Alps may report EMPTY reservation for the first
+				 * time we query it. So for the first time we
+				 * need to poll for alps_confirm_empty_timeout
+				 * time in hopes of letting the Cray race 
+				 * condition work itself out.  Once we hit the 
+				 * timeout we assume that there are no ALPS
+				 * claims (i.e. apruns) on the reservation and
+				 * the suspend can proceed.
+				 */
+				timeout_val = alps_confirm_empty_timeout;
+				first_status_was_empty = 1;
+			} 
+			if (!first_status_was_empty) {
+				/* The first status response wasn't EMPTY and
+				 * the status is now EMPTY.  According to
+				 * Cray this means PBS can proceed
+				 * as if the switch request was successful.
+				 */
+				break;
+			}
+		} else {
+			/* Reset the timeout_val if we get anything besides
+			 * "EMPTY" 
+			 */
+			timeout_val = alps_confirm_switch_timeout;
+		}
+		/* Getting a transient error, sleep then retry */
+		if (first_sleep) {
+			usleep(ALPS_SWITCH_SLEEP_USECS_LONG);
+			first_sleep = 0;
+		} else {
+			usleep(ALPS_SWITCH_SLEEP_USECS_SHORT);
+		}
+		end_time = time(NULL);
+		first_status = 0;
+	}
+	if (rc == 1) {
+		sprintf(log_buffer, "Timed out after %d attempts over %ld "
+			"seconds of attempting to confirm the ALPS "
+			"SWITCH completed.", i, total_time);
+		return (PBSE_ALPS_SWITCH_ERR);
+	} else if ((rc == 2) && first_status_was_empty) {
+		sprintf(log_buffer, "Timed out after %d attempts over "
+			"%ld seconds of waiting for a status of EMPTY "
+			"to change.  Proceeding as if the SWITCH "
+			"succeeded.", i, total_time);
+		log_event(PBSEVENT_DEBUG2, PBS_EVENTCLASS_JOB, 
+			LOG_DEBUG, pjob->ji_qs.ji_jobid, log_buffer);
+		return (PBSE_NONE);
+	} else if (rc < 0) {
+		sprintf(log_buffer, "Fatal ALPS QUERY of status.");
+		return (PBSE_ALPS_SWITCH_ERR);
+	} else {
+		/* the SWITCH has completed successfully */
+		sprintf(log_buffer, "The SWITCH was confirmed after a total "
+			"of %d attempts, and %ld seconds.", i, total_time);
+		log_event(PBSEVENT_DEBUG3, PBS_EVENTCLASS_JOB, LOG_DEBUG,
+				pjob->ji_qs.ji_jobid, log_buffer);
+	}
+
+	return PBSE_NONE;
+}
+#endif /* MOM_ALPS */
+
+
 /**
  * @brief
  *	responsible for suspend/resume job.
  *
  * @param[in] pjob - pointer to job
  * @param[in] which - indication for whether suspend/resume
+ *		0 - resume
+ *		1 - suspend
```

Comment:
Perhaps we should create 2 #define for these?  Then we can use them in the code instead of always having to check what does 0 mean?

Reply:
Added